crates/polars-stream/src/nodes/parquet_source.rs

```diff
@@ -0,0 +1,1601 @@
+        }
+
+        fn update_state(&mut self, recv: &mut [PortState], send: &mut [PortState]) {
+            use std::sync::atomic::Ordering;
+
+            debug_assert!(recv.is_empty());
+            debug_assert_eq!(send.len(), 1);
+
+            if self.is_finished.load(Ordering::Relaxed) {
+                send[0] = PortState::Done;
+                debug_assert!(
+                    self.row_group_data_rx_handles
+                        .try_write()
+                        .expect("exclusive acquire should succeed at this point")
+                        .is_none(),
+                    "should have already been shut down"
+                );
+            } else if send[0] == PortState::Done {
+                {
+                    // Early shutdown - our port state was set to `Done` by the downstream nodes. This
+                    // also means we can discard any potential errors from the `shutdown()`. Note that
+                    // `shutdown()` internally unwraps any critical errors.
+                    let _ = polars_io::pl_async::get_runtime()
+                        .block_on(self.try_shutdown())
+                        .unwrap();
+                };
+                self.is_finished.store(true, Ordering::Relaxed);
+            } else {
+                send[0] = PortState::Ready
+            }
+        }
+
+        fn spawn<'env, 's>(
+            &'env mut self,
+            scope: &'s TaskScope<'s, 'env>,
+            recv: &mut [Option<RecvPort<'_>>],
+            send: &mut [Option<SendPort<'_>>],
+            _state: &'s ExecutionState,
+            join_handles: &mut Vec<JoinHandle<PolarsResult<()>>>,
+        ) {
+            use std::sync::atomic::Ordering;
+
+            debug_assert!(recv.is_empty());
+            debug_assert_eq!(send.len(), 1);
+
+            let morsel_senders = send[0].take().unwrap().parallel();
+
+            #[cfg(debug_assertions)]
+            {
+                let guard = self.row_group_data_rx_handles.try_read().unwrap();
+                let (row_group_data_receivers, _) = guard.as_ref().unwrap();
+                assert_eq!(row_group_data_receivers.len(), morsel_senders.len());
+            }
+
+            self.row_group_data_task_init_permit.take();
+            let row_group_decoder = self.row_group_decoder.as_ref().unwrap().clone();
+
+            for (i, mut morsel_tx) in morsel_senders.into_iter().enumerate() {
+                let this = &*self;
+                let row_group_decoder = row_group_decoder.clone();
+                let row_group_data_rx_handles_guard =
+                    this.row_group_data_rx_handles.try_read().unwrap();
+
+                join_handles.push(scope.spawn_task(TaskPriority::Low, async move {
+                    let (row_group_data_receivers, _) =
+                        row_group_data_rx_handles_guard.as_ref().unwrap();
+                    let row_group_data_rx = &row_group_data_receivers[i];
```

Comment:
@orlp Updated to use the distributor channel + parallel sender here

Reply:
👍